### PR TITLE
Revolution Major Victory/Failure

### PR DESCRIPTION
**Pull request by ExcessiveUseOfCobblestone**
:cl: Cobby (Idea from Cedarbridge)
add: Revolution now rewards individuals at the end of round text with a Major Victory(revs)/Failure(heads) if casualties are kept to a minimum.
/:cl:

While I understand rev is typically TDM: The mode, I thought it would be interesting to see how people changed their playstyle if they knew they'd be rewarded for keeping as many people as they can alive in the round.

This does nothing in terms of the actual mechanics of rev, but the idea is that people are desiring of getting the best greentext they can, and targeting specifically heads/headrevs will give you a "higher score" than targeting everyone with/without an implant.

I chose 50 (percent) because this leaves a considerable amount of wiggle room for individuals on both ends to kill when needed. If it's too low I can just edit as needed but i'd rather start low and push up rather than start high and people feel like they can't kill anyone but heads/headrevs. NOTE: For a 20pop rev to major win, only 5 individuals may die outside of the heads assuming all head positions are filled.

Also added a gamemode helper should you ever want to get the amount of alive people for a gamemode or check for a percentage of people alive 😄. 

Thank Cedarbridge for the idea.

### CODE REVIEW

**Comment by WJohn** [comment]
Sure. Sounds harmless enough.

**Comment by kevinz000** [comment]
WOOT I CAN FINALLY GET BETTER GREENTEXT WHEN I POWERGAME REVS!

**Comment by ExcessiveUseOfCobblestone** [comment]
DNM because I need to add a check for suicides

**Comment by AnonymousNow** [comment]
Sounds like a variation of the "major rev victory if all the heads are marooned alive" suggestion I've been banging on about for two or three years.

**Comment by Qbopper** [comment]
would be pretty funny to see revs hogtie all the heads and throw them into the mining shuttle

**Comment by PKPenguin321** [comment]
I guess it doesn't matter since it's just flavor text but something just doesn't feel right about encouraging non violence in the GLORIOUS REVOLUTION BATHED IN BLOOD mode

**Comment by ExcessiveUseOfCobblestone** [comment]
You're revolting against the heads, not from everyone who isn't flashed.

File: code/game/gamemodes/game_mode.dm [hunk 1]
```diff
@@ -533,4 +533,22 @@
 //By default nuke just ends the round
 /datum/game_mode/proc/OnNukeExplosion(off_station)
 	if(off_station < 2)
-		station_was_nuked = TRUE //Will end the round on next check.
\ No newline at end of file
+		station_was_nuked = TRUE //Will end the round on next check.
+
+/datum/game_mode/proc/crew_live_count(var/min_percent_as_num) //return number of survivors OR T/F if argument is set
```
Comment (Cyberboss) [inline code/game/gamemodes/game_mode.dm]: remove var/

File: code/game/gamemodes/game_mode.dm [hunk 2]
```diff
@@ -533,4 +533,22 @@
...
 /datum/game_mode/proc/OnNukeExplosion(off_station)
 	if(off_station < 2)
-		station_was_nuked = TRUE //Will end the round on next check.
\ No newline at end of file
+		station_was_nuked = TRUE //Will end the round on next check.
+
+/datum/game_mode/proc/crew_live_count(var/min_percent_as_num) //return number of survivors OR T/F if argument is set
+	var/players = 0
+	var/survivors = 0
+	for(var/i in GLOB.mob_list)
```
Comment (Cyberboss) [inline code/game/gamemodes/game_mode.dm]: use alive_mob_list or whatever @vuonojenmustaturska named it

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by optimumtact** [comment]
Won't do jack other than slyly bypassing policy to code in you absurd SOS era expectation of Blartsec.

Oh we gave them more objectives! That'll control em!
Just like how changelings politely do their laundry list of traitor objectives and don't murder everyone they can with their absurd endurance and built in sword.

Rev is a simple and functional mode and the pure violence of it is one of the high points, such a shame you two can't realise that.

File: code/game/gamemodes/game_mode.dm (hunk 2 above)
Comment (ExcessiveUseOfCobblestone) [reply code/game/gamemodes/game_mode.dm]: But i'm also counting the dead with min_percent_as_num

**Comment by PKPenguin321** [comment]
If you're standard crew in a rev round and you're not going out and fighting the revs with your own two hands you're playing the mode wrong

**Comment by FantasticFwoosh** [comment]
Captain & heads suicide naturally over the course of the round or get convinced by crew to all go out on a mission together.

>> Major Victory! Bloodless revolution!

**Comment by KorPhaeron** [comment]
@AnturK would this conflict with your stuff?

**Comment by AnturK** [comment]
Yes, if you could wait with this i'd be grateful

**Comment by ExcessiveUseOfCobblestone** [comment]
😭

**Comment by octareenroon91** [comment]
Can extra-high survival rate allow the round to continue?

**Comment by ExcessiveUseOfCobblestone** [comment]
did whatever whatever get done so I can resolve conflicts now @AnturK ?

**Comment by AnturK** [comment]
Yes should be good, just take notice that results go to separate function now.

**Comment by ExcessiveUseOfCobblestone** [comment]
?!!!!!

**Comment by ExcessiveUseOfCobblestone** [comment]
ok there we go

**Comment by ExcessiveUseOfCobblestone** [comment]
You can remove the DNM now :/

File: code/game/gamemodes/revolution/revolution.dm [hunk 3]
```diff
@@ -173,18 +173,35 @@
...
 	else if(finished == 2)
-		SSticker.mode_result = "loss - rev heads killed"
-		SSticker.news_report = REVS_LOSE
+		if(crew_live_count(50))
+			SSticker.mode_result = "loss - revheads killed"
+			SSticker.news_report = REVS_LOSE
+		else
+			SSticker.mode_result = "halfwin - revheads killed - too many died"
+      SSticker.news_report = REVS_LOSE_COSTLY
+	..()
```
Comment (AnturK) [inline code/game/gamemodes/revolution/revolution.dm]: No need to call ..() twice

File: code/game/gamemodes/revolution/revolution.dm [hunk 4]
```diff
@@ -173,18 +173,35 @@
...
+	return TRUE
 
 //TODO What should be displayed for revs in non-rev rounds
 /datum/game_mode/revolution/special_report()
 	if(finished == 1)
-		return "<span class='redtext big'>The heads of staff were killed or exiled! The revolutionaries win!</span>"
+    if(crew_live_count(50))
+		  return "<FONT size = 3><B>Revolution Major Success</B></FONT></br><span class='redtext'>The heads of staff were killed or exiled! The revolution was a great success!</span>"
```
Comment (AnturK) [inline code/game/gamemodes/revolution/revolution.dm]: Please use the spans and appropriate classes

**Review by AnturK — CHANGES_REQUESTED** [review]

File: code/game/gamemodes/revolution/revolution.dm (hunk 4 above)
Comment (ExcessiveUseOfCobblestone) [reply code/game/gamemodes/revolution/revolution.dm]: I stole the format from nukeops unless you recently changed them to give them actual spans.

**Comment by octareenroon91** [comment]
>mulligan sounds odd on rev.

A bloodless coup opens the way to a nice round of Extended

**Comment by Bawhoppen** [comment]
People need to stop suggesting bloating PRs with stuff that belongs in their own PR.

**Comment by lzimann** [comment]
@ExcessiveUseOfCobblestone travis

**Comment by octareenroon91** [comment]
You have whitespace problems.

````
code/game/gamemodes/revolution/revolution.dm:178:error: inconsistent indentation
code/game/gamemodes/revolution/revolution.dm:188:error: inconsistent indentation
code/game/gamemodes/revolution/revolution.dm:194:error: inconsistent indentation
code/game/gamemodes/revolution/revolution.dm:195:error: inconsistent indentation
code/game/gamemodes/revolution/revolution.dm:196:error: inconsistent indentation
code/game/gamemodes/revolution/revolution.dm:197:error: inconsistent indentation
code/game/gamemodes/revolution/revolution.dm:196:warning: empty 'else' clause
code/game/gamemodes/revolution/revolution.dm:200:error: inconsistent indentation
code/game/gamemodes/revolution/revolution.dm:201:error: inconsistent indentation
tgstation.mdme.dmb - 8 errors, 1 warning (12/12/17 1:09 am)
````

**Comment by Robustin** [comment]
50% doesn't mean anything. Any revolution that has already slipped to 40 or even 30% casualties has long stopped caring about keeping a low body count. 

You took the opposite approach necessary to make this meaningful. 

Make the casualty requirement rigorous and add a special ending if either side wins without killing the other side's heads (bonus points if you can acquire art representing the glory of a bloodless victory) and you'll see a real effort to avoid the usual deathmatch. 

Also it strikes me as unfair to the revs since Command starts with access to several non-lethal ways to defeat a revolution but revs basically need a lowpop round or a miracle to neutralize heads AND security without bloodshed. The numbers for either side shouldn't be symmetrical.

**Finally: Why am I the first to notice that your "COSTLY REV VICTORY" has the "COSTLY CREW VICTORY" message? It literally says the rev was defeated.**

File: code/game/gamemodes/revolution/revolution.dm [hunk 5]
```diff
@@ -173,18 +173,34 @@
...
+	return TRUE
 
 //TODO What should be displayed for revs in non-rev rounds
 /datum/game_mode/revolution/special_report()
 	if(finished == 1)
-		return "<span class='redtext big'>The heads of staff were killed or exiled! The revolutionaries win!</span>"
+    	if(crew_live_count(50))
+			return "<FONT size = 3><B>Revolution Major Success</B></FONT></br><span class='redtext'>The heads of staff were killed or exiled! The revolution was a great success!</span>"
+    	else
+			return "<FONT size = 3><B>Revolution Pyrrhic Victory</B></FONT></br><span class='redtext'>The crew successfully defended against the revolution but too many lives were lost in the process.</span>"
```
Comment (Robustin) [inline code/game/gamemodes/revolution/revolution.dm]: This is an exact copy of the "Staff Pyrrhic Victory" message - it states the the Revolution was defeated.

**Comment by lzimann** [comment]
ping whenever you come back to this

**Pull request closed without merging**